src/series/algorithm/waterfall.js

```diff
@@ -0,0 +1,107 @@
+import d3 from 'd3';
+import {noop} from '../../util/fn';
+
+export default function() {
+
+    var xValueKey = '',
+        yValue = function(d) { return d.y; },
+        startsWithTotal = false,
+        totals = noop;
+
+    var waterfall = function(data) {
+        var length = data.length,
+            i = 0,
+            previousEnd = yValue(data[0]),
+            start,
+            end,
+            total,
+            result = [],
+            finalIndex = length - 1,
+            customTotals = false;
+
+        if (startsWithTotal) {
+            // First value is a total
+            result.push({
+                domain: data[0][xValueKey],
+                start: 0,
+                end: previousEnd,
+                direction: 0
+            });
+            i = 1;
+        }
+
+        for (i; i < length; i += 1) {
+            start = previousEnd;
+            end = yValue(data[i]);
+
+            result.push({
+                domain: data[i][xValueKey],
+                start: start,
+                end: end,
+                direction: end - start > 0 ? 1 : -1
+            });
+
+            // don't have two totals at the end
+            total = totals(data[i], i, data);
+            if (total) {
+                customTotals = true;
+
+                // Insert a total value here
+                result.push({
+                    domain: total,
+                    start: 0,
+                    end: end,
+                    direction: 0
+                });
+            }
+
+            previousEnd = end;
+        }
+
+        if (!customTotals) {
```

Comment:
I'm pretty sure you could remove this logic by having `totals` default to a function that generates a total for the last data point.